File: src/org/genyris/core/Dictionary.java
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.core;

import java.util.Calendar;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.TreeMap;

import org.genyris.exception.AccessException;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.DynamicEnvironment;
import org.genyris.interp.Environment;
import org.genyris.interp.UnboundException;

public class Dictionary extends Atom implements Environment {
	private Map _dict;

	protected Environment _parent;

    private static Map mapFactory() {
    	return new HashMap();
    }

	public Dictionary() {
		_dict = mapFactory();
		_parent = null;
	}

	public Dictionary(Environment parent) {
		_dict = mapFactory();
		_parent = parent;
	}

	public Dictionary(SimpleSymbol key, Exp value, Environment parent) {
		_dict = mapFactory();
		_dict.put(key, value);
		_parent = parent;
	}

	public Dictionary(Environment parent, final HashMap map) {
		_dict = new HashMap(map);
		_parent = parent;
	}

	public Environment getParent() {
		return _parent;
	}

	public void acceptVisitor(Visitor guest) throws GenyrisException {
		guest.visitDictionary(this);
	}

	public boolean hasKey(Exp a) {
		return _dict.containsKey(a);
	}

	public Exp asAlist() {
		Map keys = new TreeMap(_dict);
		Iterator iter = keys.keySet().iterator();
		// TODO Sort the keyset to get a consistent result for test cases.
		Exp result = _parent.getNil();
		while (iter.hasNext()) {
			Exp key = (Exp)iter.next();
			Exp value = (Exp)  _dict.get(key);
			Exp tmp = new PairEquals( new DynamicSymbol((SimpleSymbol)key), value);
			result = new Pair(tmp, result);
		}
		return new Pair(_parent.getSymbolTable().DICT(), result);
	}

	public void defineVariableRaw(SimpleSymbol sym, Exp valu) throws GenyrisException {
		if (sym == CLASSES()) {
			setClasses(valu, _parent.getNil());
			return;
		} else {
			_dict.put(sym, valu);
			return;
		}
	}

	public void defineDynamicVariable(DynamicSymbol sym, Exp valu) throws GenyrisException {
		defineVariableRaw(sym.getRealSymbol(), valu);
	}

	private Exp CLASSES() {
		return _parent.getSymbolTable().CLASSES();
	}

	private Exp SELF() {
		return _parent.getSymbolTable().SELF();
	}

	private Exp SUPERCLASSES() {
		return _parent.getSymbolTable().SUPERCLASSES();
	}

	private Exp VARS() {
		return _parent.getSymbolTable().VARS();
	}

	private Exp getVarsList() {
		Iterator iter = _dict.keySet().iterator();
		Exp result = new Pair(VARS(), _parent.getNil());
		while (iter.hasNext()) {
			Exp key = (Exp) iter.next();
			result = new Pair(key, result);
		}
		return result;
	}

	private Exp lookupInClasses(DynamicSymbol symbol) throws UnboundException {
		Exp classes = getClasses(_parent);
		while (classes != _parent.getNil()) {
			try {
				if(!(classes.car() instanceof Environment)) {
					throw new UnboundException("damaged class in lass list: " + classes.car());
				}
				Environment klass = (Environment) (classes.car());
				try {
					return (Exp) klass.lookupInThisClassAndSuperClasses(symbol);
				} catch (UnboundException e) {
					;
				} finally {
					classes = classes.cdr();
				}
			} catch (AccessException e) {
				throw new UnboundException("bad classes list in object");
			}
		}
		throw new UnboundException("key " + symbol.toString() + 
				" not found in dict: " + this);
	}

	public Exp lookupInThisClassAndSuperClasses(DynamicSymbol symbol)
			throws UnboundException {
		if (_dict.containsKey(symbol.getRealSymbol())) {
			return (Exp) _dict.get(symbol.getRealSymbol());
		} else {
			return lookupInSuperClasses(symbol);
		}
	}

	private Exp lookupInSuperClasses(DynamicSymbol symbol) throws UnboundException {
		if (!_dict.containsKey(SUPERCLASSES())) {
			throw new UnboundException("object has no superclasses");
		}
		Exp superclasses = (Exp) _dict.get(SUPERCLASSES());
		while (superclasses != _parent.getNil()) {
			try {
				Environment klass = (Environment) (superclasses.car());
				try {
					return (Exp) klass.lookupInThisClassAndSuperClasses(symbol);
				} catch (UnboundException e) {
					;
				} finally {
					superclasses = superclasses.cdr();
				}
			} catch (AccessException e) {
				throw new UnboundException("bad classes list in object");
			}
		}
		throw new UnboundException("key " + symbol.toString() + 
				" not found in dict: " + this);
	}

	public void setVariableValue(Symbol symbol, Exp valu) throws UnboundException {
		symbol.setVariableValue(this, valu);
	}
	public void setDynamicVariableValueRaw(SimpleSymbol sym, Exp valu) throws UnboundException {
		if (sym == CLASSES()) {
			try {
				setClasses(valu, _parent.getNil());
			} catch (AccessException ignore) {
			}
		} else {
			if (_dict.containsKey(sym)) {
				_dict.put(sym, valu);
			} else {
				throw new UnboundException("in object, undefined variable: " + sym);
			}
		}
	}
	public void setDynamicVariableValue(DynamicSymbol symbol, Exp valu) throws UnboundException {
		setDynamicVariableValueRaw(symbol.getRealSymbol(), valu);
	}
	
	public void setLexicalVariableValue(SimpleSymbol symbol, Exp valu) throws UnboundException {
		throw new UnboundException("Dictionary expected dynamic symbol ref " 	+ symbol);
	}

	public String toString() {
		return "<dict " + asAlist().toString() + ">";
	}

	public Exp lookupVariableShallow(SimpleSymbol symbol) throws UnboundException {
		if (symbol == CLASSES()) {
			return getClasses(_parent);
		} else if (_dict.containsKey(symbol)) {
			return (Exp) _dict.get(symbol);
		} else {
			throw new UnboundException("key " + symbol.toString() + 
					" not found in dict: " + this);
		}
	}

	public SimpleSymbol getNil() {
		return _parent.getNil();
	}

	public Symbol internString(String symbolName) {
		return _parent.internString(symbolName);
	}

	public Symbol getBuiltinClassSymbol(Internable table) {
		return table.DICTIONARY();
	}

	public Exp getSelf() {
		return this;
	}
	public Internable getSymbolTable() {
		return _parent.getSymbolTable();
	}

	public Exp eval(Environment env) throws GenyrisException {
		return this;
	}

	public Exp lookupVariableValue(Symbol symbol) throws UnboundException {
		try {
			return symbol.lookupVariableValue(this);
		} catch (UnboundException e) {
			throw new UnboundException("variable " + symbol + " unbound in " + this);
		}
	}
	public Exp lookupDynamicVariableValue(DynamicSymbol dsymbol) throws UnboundException {
		Symbol symbol = ((DynamicSymbol)dsymbol).getRealSymbol();

		if (symbol == SELF()) {
			return this;
		} else if (symbol == CLASSES()) {
			return getClasses(_parent);
		} else if (symbol == VARS()) {
			return getVarsList();
		} else if (_dict.containsKey(symbol)) {
			return (Exp) _dict.get(symbol);
		}
		try {
			return lookupInClasses(dsymbol);
		} catch (UnboundException ignore) {
		}

		if (_dict.containsKey(SUPERCLASSES())) {
			return lookupInSuperClasses(dsymbol);
		}
		throw new UnboundException("unbound " + symbol.toString());
	}

	public Exp lookupLexicalVariableValue(SimpleSymbol symbol) throws UnboundException {
		throw new UnboundException("no lexical symbols in Dictionary: " + this);
	}

	public void defineLexicalVariable(SimpleSymbol symbol, Exp valu) throws GenyrisException {
		throw new UnboundException("no lexical symbols in Dictionary: " + symbol);		
	}

	public void defineVariable(Symbol symbol, Exp valu) throws GenyrisException {
			symbol.defineVariable(this, valu);
	}
	public Environment  makeEnvironment(Environment parent) throws GenyrisException {
		Map bindings = mapFactory();
		bindings.put(SELF(), this);
		return new DynamicEnvironment(parent, bindings, this);
	}
	public Dictionary addProperty(Environment env, String name, Exp value) throws GenyrisException {
    	defineDynamicVariable(new DynamicSymbol((SimpleSymbol) env.internString(name)), 
    			value);
		return this;
	}

	public void defineInt(String name, int i) throws GenyrisException {
		defineDynamicVariable(new DynamicSymbol((SimpleSymbol) internString(name)), new Bignum(i));	
	}

	public void defineString(String name, String string2) throws GenyrisException {
		defineDynamicVariable(new DynamicSymbol((SimpleSymbol) internString(name)), new StrinG(string2));	
	}
	public void defineSymbol(String name, String string2) throws GenyrisException {
		defineDynamicVariable(new DynamicSymbol((SimpleSymbol) internString(name)), internString(string2));	
	}
}

File: src/org/genyris/core/Dictionary.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.core

import org.genyris.exception.AccessException
import org.genyris.exception.GenyrisException
import org.genyris.interp.DynamicEnvironment
import org.genyris.interp.Environment
import org.genyris.interp.UnboundException
import java.util.*

open class Dictionary : Atom, Environment {
    var map: MutableMap<*, *>
        protected set

    var parent: Environment?
        protected set

    constructor() {
        this.map = mapFactory()
        this.parent = null
    }

    constructor(parent: Environment?) {
        this.map = mapFactory()
        this.parent = parent
    }

    constructor(key: Symbol?, value: Exp?, parent: Environment?) {
        this.map = mapFactory()
        map.put(key, value)
        this.parent = parent
    }

    constructor(parent: Environment?, map: HashMap<*, *>) {
        this.map = HashMap<Any?, Any?>(map)
        this.parent = parent
    }

    @kotlin.Throws(GenyrisException::class)
    override fun acceptVisitor(guest: Visitor) {
        guest.visitDictionary(this)
    }

    fun hasKey(a: Exp?): Boolean {
        return map.containsKey(a)
    }

    fun asAlist(): Exp {
        val NIL: Exp? = parent!!.getNil()
        val keys: SortedSet<*> = TreeSet<Any?>(map.keySet())
        val iter: MutableIterator<Exp?> = keys.iterator()
        var result = NIL
        var tail = result
        while (iter.hasNext()) {
            val key = iter.next()
            val value = map.get(key) as Exp?
            val association: Pair = PairEquals(DynamicSymbol(key as SimpleSymbol?), value)
            val newEnd = Pair(association, NIL) // TODO Add an efficient append() function for conses.
            if (tail === NIL) {
                tail = newEnd
                result = tail
            } else {
                (tail as Pair).setCdr(newEnd)
                tail = newEnd
            }
        }
        return Pair(parent!!.getSymbolTable().DICT(), result)
    }

    @kotlin.Throws(GenyrisException::class)
    fun defineVariableRaw(sym: Symbol?, valu: Exp?) {
        if (sym === CLASSES()) {
            setClasses(valu, parent!!.getNil())
        } else {
            map.put(sym, valu)
        }
    }

    @kotlin.Throws(GenyrisException::class)
    override fun defineDynamicVariable(sym: DynamicSymbol, valu: Exp?) {
        defineVariableRaw(sym.getRealSymbol(), valu)
    }

    protected fun CLASSES(): Exp? {
        return parent!!.getSymbolTable().CLASSES()
    }

    protected fun SELF(): Exp? {
        return parent!!.getSymbolTable().SELF()
    }

    private fun SUPERCLASSES(): Exp? {
        return parent!!.getSymbolTable().SUPERCLASSES()
    }

    private fun VARS(): Exp? {
        return parent!!.getSymbolTable().VARS()
    }

    override fun dir(table: Internable): Exp {
        val iter: MutableIterator<*> = map.keySet().iterator()
        var result: Exp = Pair.Companion.cons3(
            DynamicSymbol(table.SELF()),
            DynamicSymbol(table.VARS()),
            DynamicSymbol(table.CLASSES()), table.NIL()
        )
        while (iter.hasNext()) {
            val key: Exp = DynamicSymbol(iter.next() as SimpleSymbol?)
            result = Pair(key, result)
        }
        return result
    }

    @kotlin.Throws(UnboundException::class)
    private fun lookupInClasses(symbol: DynamicSymbol): Exp? {
        var classes = getClasses(this.parent)
        while (classes !== parent!!.getNil()) {
            try {
                if (classes.car() !is Environment) {
                    throw UnboundException("damaged class in lass list: " + classes.car())
                }
                val klass = (classes.car()) as Environment
                try {
                    return klass.lookupInThisClassAndSuperClasses(symbol)
                } catch (e: UnboundException) {
                } finally {
                    classes = classes.cdr()
                }
            } catch (e: AccessException) {
                throw UnboundException("bad classes list in object")
            }
        }
        throw UnboundException(
            "key " + symbol.toString() +
                    " not found in dict: " + this
        )
    }

    @kotlin.Throws(UnboundException::class)
    override fun lookupInThisClassAndSuperClasses(symbol: DynamicSymbol): Exp? {
        if (map.containsKey(symbol.getRealSymbol())) {
            return map.get(symbol.getRealSymbol()) as Exp?
        } else {
            return lookupInSuperClasses(symbol)
        }
    }

    @kotlin.Throws(UnboundException::class)
    private fun lookupInSuperClasses(symbol: DynamicSymbol): Exp? {
        if (!map.containsKey(SUPERCLASSES())) {
            throw UnboundException("object has no superclasses")
        }
        var superclasses = map.get(SUPERCLASSES()) as Exp
        while (superclasses !== parent!!.getNil()) {
            try {
                val klass = (superclasses.car()) as Environment
                try {
                    return klass.lookupInThisClassAndSuperClasses(symbol)
                } catch (e: UnboundException) {
                } finally {
                    superclasses = superclasses.cdr()
                }
            } catch (e: AccessException) {
                throw UnboundException("bad classes list in object")
            }
        }
        throw UnboundException(
            "key " + symbol.toString() +
                    " not found in dict: " + this
        )
    }

    @kotlin.Throws(UnboundException::class)
    override fun setVariableValue(symbol: Symbol, valu: Exp?) {
        symbol.setVariableValue(this, valu)
    }

    @kotlin.Throws(UnboundException::class)
    open fun setDynamicVariableValueRaw(sym: Symbol?, valu: Exp?) {
        if (sym === CLASSES()) {
            try {
                setClasses(valu, parent!!.getNil())
            } catch (ignore: AccessException) {
            }
        } else {
            if (map.containsKey(sym)) {
                map.put(sym, valu)
            } else {
                throw UnboundException("in object, undefined variable: " + sym)
            }
        }
    }

    @kotlin.Throws(UnboundException::class)
    override fun setDynamicVariableValue(symbol: DynamicSymbol, valu: Exp?) {
        setDynamicVariableValueRaw(symbol.getRealSymbol(), valu)
    }

    @kotlin.Throws(UnboundException::class)
    override fun setLexicalVariableValue(symbol: SimpleSymbol?, valu: Exp?) {
        throw UnboundException("Dictionary expected dynamic symbol ref " + symbol)
    }

    override fun toString(): String {
        return "<dict " + asAlist().toString() + ">"
    }

    @kotlin.Throws(UnboundException::class)
    fun lookupVariableShallow(symbol: SimpleSymbol): Exp? {
        if (symbol === CLASSES()) {
            return getClasses(this.parent)
        } else if (map.containsKey(symbol)) {
            return map.get(symbol) as Exp?
        } else {
            throw UnboundException(
                "key " + symbol.toString() +
                        " not found in dict: " + this
            )
        }
    }

    override fun getNil(): SimpleSymbol? {
        return parent!!.getNil()
    }

    override fun internString(symbolName: String?): Symbol? {
        return parent!!.internString(symbolName)
    }

    override fun getBuiltinClassSymbol(table: Internable): Symbol? {
        return table.DICTIONARY()
    }

    override fun getSelf(): Exp {
        return this
    }

    override fun getSymbolTable(): Internable? {
        return parent!!.getSymbolTable()
    }

    @kotlin.Throws(GenyrisException::class)
    override fun eval(env: Environment?): Exp {
        return this
    }

    @kotlin.Throws(UnboundException::class)
    override fun lookupVariableValue(symbol: Symbol): Exp? {
        try {
            return symbol.lookupVariableValue(this)
        } catch (e: UnboundException) {
            throw UnboundException("variable " + symbol + " unbound in " + this)
        }
    }

    @kotlin.Throws(UnboundException::class)
    override fun lookupDynamicVariableValue(dsymbol: DynamicSymbol): Exp? {
        val symbol = dsymbol.getRealSymbol()

        if (symbol === SELF()) {
            return this
        } else if (symbol === CLASSES()) {
            return getClasses(this.parent)
        } else if (symbol === VARS()) {
            return dir(parent!!.getSymbolTable())
        } else if (map.containsKey(symbol)) {
            return map.get(symbol) as Exp?
        }
        try {
            return lookupInClasses(dsymbol)
        } catch (ignore: UnboundException) {
        }

        if (map.containsKey(SUPERCLASSES())) {
            return lookupInSuperClasses(dsymbol)
        }
        throw UnboundException("unbound " + symbol.toString())
    }

    @kotlin.Throws(UnboundException::class)
    override fun lookupLexicalVariableValue(symbol: SimpleSymbol?): Exp? {
        throw UnboundException("no lexical symbols in Dictionary: " + this)
    }

    @kotlin.Throws(GenyrisException::class)
    override fun defineLexicalVariable(symbol: SimpleSymbol?, valu: Exp?) {
        throw UnboundException("no lexical symbols in Dictionary: " + symbol)
    }

    @kotlin.Throws(GenyrisException::class)
    override fun defineVariable(symbol: Symbol, valu: Exp?) {
        symbol.defineVariable(this, valu)
    }

    @kotlin.Throws(GenyrisException::class)
    override fun makeEnvironment(parent: Environment?): Environment {
        val bindings: MutableMap<*, *> = mapFactory()
        bindings.put(SELF(), this)
        return DynamicEnvironment(parent, bindings, this)
    }

    @kotlin.Throws(GenyrisException::class)
    fun addProperty(env: Environment, name: String?, value: Exp?): Dictionary {
        defineDynamicVariable(
            DynamicSymbol(env.internString(name)),
            value
        )
        return this
    }

    @kotlin.Throws(GenyrisException::class)
    fun defineInt(name: String?, i: Int) {
        defineDynamicVariable(DynamicSymbol(internString(name)), Bignum(i))
    }

    @kotlin.Throws(GenyrisException::class)
    fun defineString(name: String?, string2: String?) {
        defineDynamicVariable(DynamicSymbol(internString(name)), StrinG(string2))
    }

    @kotlin.Throws(GenyrisException::class)
    fun defineSymbol(name: String?, string2: String?) {
        defineDynamicVariable(DynamicSymbol(internString(name)), internString(string2))
    }

    override fun isBound(s: Symbol?): Boolean {
        return hasKey(s)
    }

    override fun compareTo(o: Any?): Int {
        return if (this === o) 0 else 1
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || getClass() != o.getClass()) {
            return false
        }
        return this.map == (o as Dictionary).map
    }

    companion object {
        protected fun mapFactory(): MutableMap<*, *> {
            return HashMap<Any?, Any?>()
        }
    }
}
